R/rectangle.R

```diff
@@ -167,104 +222,461 @@ hoist <- function(.data, .col, ..., .remove = TRUE, .simplify = TRUE, .ptype = l
+    return(x)
+  }
+
+  index <- indices[[1L]]
+  indices <- indices[-1L]
+
+  is_valid_index <-
+    (is.numeric(index) && (index <= vec_size(x))) ||
+    (is.character(index) && has_name(x, index))
+
+  if (!is_valid_index) {
+    # Nothing to do if the `pluck()` missed entirely
+    return(x)
+  }
+
+  if (n_indices == 1L) {
+    # At base index, remove it
+    if (is.numeric(index)) {
+      x <- x[-index]
+    } else if (is.character(index)) {
+      x <- x[setdiff(names(x), index)]
+    } else {
+      abort("Internal error: Only character and numeric values are valid indices.")
+    }
+  } else {
+    # Not at base index yet, continue recursion
+    x[[index]] <- strike(x[[index]], indices)
+  }
+
+  x
+}
+
 #' @export
 #' @rdname hoist
-#' @param values_to Name of column to store vector values. Defaults to `col`.
-#' @param indices_to A string giving the name of column which will contain the
-#'   inner names or position (if not named) of the values. Defaults to `col`
-#'   with `_id` suffix
-#' @param indices_include Add an index column? Defaults to `TRUE` when `col`
-#'   has inner names.
+#' @param values_to A single string representing the column name to store the
+#'   unnested values in. If multiple columns are specified in `col`, this can
+#'   also be a glue string containing `"{col}"` to provide a template for the
+#'   column names. If `NULL`, defaults to `"{col}"`.
+#' @param indices_to A string giving the name of the column which will contain
+#'   the inner names or positions (if not named) of the values. If multiple
+#'   columns are specified in `col`, this can also be a glue string containing
+#'   `"{col}"` to provide a template for the column names. If `NULL`, defaults
+#'   to `values_to` suffixed with `"_id"`.
+#' @param indices_include A single logical value specifying whether or not to
+#'   add an index column. If any value has inner names, the index column will be
+#'   a character vector of those names, otherwise it will be an integer vector
+#'   of positions. If `NULL`, defaults to `TRUE` if any value has inner names
+#'   or if `indices_to` is provided.
+#'
+#'   If `indices_to` is provided, then `indices_include` can't be `FALSE`.
 #' @inheritParams unnest
-unnest_longer <- function(data, col,
+unnest_longer <- function(data,
+                          col,
                           values_to = NULL,
                           indices_to = NULL,
                           indices_include = NULL,
                           names_repair = "check_unique",
                           simplify = TRUE,
                           ptype = list(),
-                          transform = list()
-                          ) {
-
+                          transform = list()) {
   check_present(col)
-  col <- tidyselect::vars_pull(names(data), !!enquo(col))
+  # TODO: Use `allow_rename = FALSE`.
+  # Requires https://github.com/r-lib/tidyselect/issues/225.
+  cols <- tidyselect::eval_select(enquo(col), data)
```

Comment:
For now we just rely on the assumption that the user didn't do any renaming, which should be okay for now